pkg/monitor/agent.go

```diff
@@ -29,43 +29,55 @@ import (
 type agentMonitor struct {
 	client  clientset.Interface
 	querier agentquerier.AgentQuerier
+	// agentCRD is the desired state of agent monitoring CRD which agentMonitor expects.
+	agentCRD *v1beta1.AntreaAgentInfo
 }
 
 // NewAgentMonitor creates a new agent monitor.
 func NewAgentMonitor(client clientset.Interface, querier agentquerier.AgentQuerier) *agentMonitor {
-	return &agentMonitor{client: client, querier: querier}
+	return &agentMonitor{client: client, querier: querier, agentCRD: nil}
 }
 
 // Run creates AntreaAgentInfo CRD first after controller is running.
 // Then updates AntreaAgentInfo CRD every 60 seconds.
 func (monitor *agentMonitor) Run(stopCh <-chan struct{}) {
 	klog.Info("Starting Antrea Agent Monitor")
-	agentCRD := monitor.getAgentCRD()
-	var err error = nil
 
-	// Initialize agent monitoring CRD.
-	if agentCRD == nil {
-		agentCRD, err = monitor.createAgentCRD()
+	// Initialize agent monitoring CRD util initialization succeeds.
+	wait.PollImmediateUntil(time.Minute, func() (done bool, err error) {
+		initCRD := monitor.initAgentCRD()
+		if initCRD != nil {
+			klog.Infof("Initialized agent monitoring CRD %+v", monitor.agentCRD)
+			return true, nil
+		} else {
+			return false, nil
+		}
+	}, stopCh)
```

Comment:
Maybe we don't really need a separate poll here. The logic can be in one loop:
```
if monitor.agentCRD != nil {
	if monitor.agentCRD, err := monitor.partialUpdateAgentCRD(monitor.agentCRD); err == nil {
		return
	}
}

monitor.agentCRD, err = monitor.getAgentCRD()

if errors.IsNotFound(err) {
	monitor.agentCRD, err = monitor.createAgentCRD()
	if err != nil {
		klog.Error(...)
		monitor.agentCRD = nil
	}
	return
}

if err != nil  {
	klog.Error(...)
	monitor.agentCRD = nil
	return
}

monitor.agentCRD, err := monitor.updateAgentCRD(monitor.agentCRD)
if err != nil  {
	klog.Error(...)
	monitor.agentCRD = nil
}
```

Reply:
Thanks for the suggestion. Have updated in the latest commit.